### PR TITLE
Move SPICE part from kubectl.sh to own spice.go package

### PR DESCRIPTION
**Pull request by bond95**
https://github.com/kubevirt/kubevirt/issues/181
@rmohr Please review

### CODE REVIEW

File: pkg/virtctl/spice/spice.go [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+	}
+	body, err := DownloadSpice(vm, config.Host)
+	if err != nil {
+		log.Fatalf(err.Error())
+		return 1
+	}
+	if details {
+		fmt.Printf("%s", body)
+	} else {
+		f, err := os.Create(SPICE_FILENAME)
```
Comment (rmohr) [inline pkg/virtctl/spice/spice.go]: Maybe we can improve that a little bit and write it into a temp file which we delete after we are done:

```golang
file, err := ioutil.TempFile(os.TempDir(), "console.vv")
defer os.Remove(file.Name())
```

File: pkg/virtctl/spice/spice.go [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+const FLAG = "spice"
+const SPICE_FILENAME = "console.vv"
+const VM_SPICE_URL = "http://%s/apis/kubevirt.io/v1alpha1/namespaces/default/vms/%s/spice"
+
+type Spice struct {
+}
+
+func DownloadSpice(vm, host string) (string, error) {
+	var path string = fmt.Sprintf(VM_SPICE_URL, host, vm)
+	req, err := http.NewRequest("GET", path, nil)
```
Comment (rmohr) [inline pkg/virtctl/spice/spice.go]: We can use the rest client from kubernetes here:


```golang
body, err := restClient.Get().
  Resource("vms").SetHeader("Accept", rest.MIME_INI).
  SubResource("spice").
  Namespace(kubev1.NamespaceDefault).
  Name(vm.GetObjectMeta().GetName()).Do().Raw()
```

You can get a rest client which understands the VM TPR like this:

```golang
restClient, err := kubecli.GetRESTClientFromFlags(master, kubeconfig)
```

This also allows us to reuse the authentication information from kubeconfig.

File: pkg/virtctl/spice/spice.go [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+		f.Sync()
+
+		cmnd := exec.Command("remote-viewer", "console.vv")
+		cmnd.Start()
+	}
+	return 0
+}
+
+func (o *Spice) Usage() string {
+	usage := "The following options can be passed to any command:\n\n"
```
Comment (rmohr) [inline pkg/virtctl/spice/spice.go]: Here some usage examples like in  `virtctl/console/console.go` would be nice here.

File: pkg/virtctl/spice/spice.go [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+	kubeconfig, _ := flags.GetString("kubeconfig")
+	details, _ := flags.GetBool("details")
+
+	if len(flags.Args()) != 2 {
+		log.Println("VM name is missing")
+		return 1
+	}
+	vm := flags.Arg(1)
+
+	config, err := clientcmd.BuildConfigFromFlags(server, kubeconfig)
```
Comment (rmohr) [inline pkg/virtctl/spice/spice.go]: Here you can dreate the rest client instead and directly use it.

**Comment by bond95** [comment]
@rmohr fixed comments, please review.

File: pkg/virtctl/spice/spice.go [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+	"k8s.io/client-go/rest"
+	"kubevirt.io/kubevirt/pkg/kubecli"
+	"log"
+	"os"
+	"os/exec"
+)
+
+const FLAG = "spice"
+const SPICE_FILENAME = "console.vv"
+const VM_SPICE_URL = "http://%s/apis/kubevirt.io/v1alpha1/namespaces/default/vms/%s/spice"
```
Comment (rmohr) [inline pkg/virtctl/spice/spice.go]: That one can be removed

File: pkg/virtctl/spice/spice.go [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+
+		cmnd := exec.Command("remote-viewer", f.Name())
+		cmnd.Run()
+		os.Remove(f.Name())
+	}
+	return 0
+}
+
+func (o *Spice) Usage() string {
+	usage := "Usage: ./virtctl spice vm_name [--details] [common_options]\n\n"
```
Comment (rmohr) [inline pkg/virtctl/spice/spice.go]: Could you give real examples of the usage here? For instance kubectl exec --help gives:

```
Execute a command in a container.

Examples:
  # Get output from running 'date' from pod 123456-7890, using the first container by default
  kubectl exec 123456-7890 date
  
  # Get output from running 'date' in ruby-container from pod 123456-7890
  kubectl exec 123456-7890 -c ruby-container date
  
  # Switch to raw terminal mode, sends stdin to 'bash' in ruby-container from pod 123456-7890
  # and sends stdout/stderr from 'bash' back to the client
  kubectl exec 123456-7890 -c ruby-container -i -t -- bash -il

Options:
  -c, --container='': Container name. If omitted, the first container in the pod will be chosen
  -p, --pod='': Pod name
  -i, --stdin=false: Pass stdin to the container
  -t, --tty=false: Stdin is a TTY

Usage:
  kubectl exec POD [-c CONTAINER] -- COMMAND [args...] [options]
```

Maybe

```
Fetch SPICE connections details of a VM and connect via remote-viewer

Examples:
#Show connection details of the VM myvm
virtctl spice myvm --details

# Open a SPICE connection to the VM myvm via remote-viewer
virtctl spice myvm

Options:
```
The rest will be added by o.FlagSet.FlagUsages()

Comment (rmohr) [inline pkg/virtctl/spice/spice.go]: `o.FlagSet.FlagUsages()` adds options and the usage for you

**Comment by bond95** [comment]
Fixed @rmohr please review

File: pkg/virtctl/spice/spice.go [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+			log.Fatalf("Can't write to file: %s", err.Error())
+			defer os.Remove(f.Name())
+			return 1
+		}
+
+		f.Sync()
+
+		cmnd := exec.Command("remote-viewer", f.Name())
+		err = cmnd.Run()
+		defer os.Remove(f.Name())
```
Comment (rmohr) [inline pkg/virtctl/spice/spice.go]: This defer is too late, it needs to be directly after `defer f.Close()`, the rest looks good to me :)

**Comment by bond95** [comment]
@rmohr fixed =)

**Comment by rmohr** [comment]
Ups, I suggested the wrong  defer order. It is a stack, so we need

```golang
defer os.Remove(f.Name())
defer f.Close()
```

Also see https://tour.golang.org/flowcontrol/13 for a nice example.

**Comment by bond95** [comment]
Fixed @rmohr

**Comment by rmohr** [comment]
Thx! :+1:

**Pull request merged**